### PR TITLE
added non-blocking snapshot functions

### PR DESCRIPTION
**Pull request by frank26080115**
Added a non-blocking way of taking snapshots. This makes it easier to write responsive applications that require long exposure captures.

Please add the following to the `omv.sensor.rst` file

    .. function:: sensor.snapshot_start()

       Initiates a snapshot, but returns immediately. Use `sensor.snapshot_check()` to see if the image is ready. Use `sensor.snapshot_finish()` to retrieve the image.

    .. function:: sensor.snapshot_check()

       Returns whether or not the sensor has finished the snapshot started with `sensor.snapshot_start()`. True if the image is ready. Use `sensor.snapshot_finish()` to retrieve the image.

    .. function:: sensor.snapshot_finish()

       Returns the image captured after using `sensor.snapshot_start()`. If the image is not ready, it will wait until it is (or timeout), then return the image.

I have tested this on my OpenMV H7 Plus

### CODE REVIEW

**Comment by kwagyeman** [comment]
Thanks for the PR. I have plans to make snapshot to support double buffering and N deep image buffering soon. What you've done is possible to support. However, I'd probably want the API to be worded in a particular way. I'll take a look more closely and what you've done and ask for some changes to what the double buffering API will look like.

File: src/omv/py/py_sensor.c [hunk 1]
```diff
@@ -101,6 +101,45 @@ static mp_obj_t py_sensor_snapshot(uint n_args, const mp_obj_t *args, mp_map_t *
...
+        nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_RuntimeError, "Capture Start Failed: %d", ret));
+    }
+
+    return mp_const_none;
+}
+
+static mp_obj_t py_sensor_snapshot_check()
+{
+    bool status;
+    sensor.snapshot_check(&sensor, &status);
```
Comment (kwagyeman) [inline src/omv/py/py_sensor.c:126]: Make this just return the bool. Then it's one line.

**Comment by frank26080115** [comment]
All I really did is split the original snapshot function in two, and remove the streaming callback from my variation. No change was done to the original snapshot function, so this is drop in friendly with no risk to existing code.

I know this opens up a big can of worms... like what happens if the user screws around with the camera and buffers while a capture is on-going.

Looking forward to what you decide to implement. I will continue my own project with my changes for now.

By the way, I managed to build this while on Windows, the trick is to use Ubuntu for Windows and fix the problem in https://github.com/openmv/openmv/issues/743 , I posted a solution to the problem there

**Comment by kwagyeman** [comment]
Send the doc as a PR to: https://github.com/openmv/micropython/blob/c2938c32ef4c78d5c92e861db7367a3312ef6454/docs/library/omv.sensor.rst

Thanks for the good idea for a non-blocking idea!

Double buffering is my next thing to work on after getting the Arduino Interface Library out. Max FPS and video recording will then be possible.

File: src/omv/py/py_sensor.c (hunk 1 above)
Comment (frank26080115) [reply src/omv/py/py_sensor.c]: I made commit 3f1d58d that makes these changes for you but I can't really test the stream callback

**Comment by kwagyeman** [comment]
Okay, thanks, going to sit on this for a while given the double buffering code. But, I'll see what we can do to make a non-blocking method.

**Comment by iabdalkader** [comment]
> I know this opens up a big can of worms... like what happens if the user screws around with the camera and buffers while a capture is on-going.

It does, if we merge this we have to maintain it and handle all the possible ways these functions can be used (or misused), to add to what you mentioned, we need to handle `snapshot_start` being called more than once, or `snapshot_finish` called multiple times, or before/without `snapshot_start`, or buffers allocated mid-snapshot. In its current state, it just doesn't handle any of that. The same concerns apply to modifying snapshot to be non-blocking, so I'm just saying we have to be very careful with this feature.

**Comment by kwagyeman** [comment]
This is not going to be needed in the future for an updated version of our driver. I will implement triple buffering as an option you can just turn on. When you do so the driver will generate 3 frame buffers with the camera driver delivering the latest image to 1 of the two buffers you are not working on. This will make the camera interface non-blocking and deliver you the latest image always when you want it (you'll also be technically able to run at frame rates faster than the camera can deliver images).

Not sure exactly how I'm make an image fifo, but, otherwise, it will be fine.

**Pull request closed without merging**